pkg/alibabacloud/eni/instances.go

```diff
@@ -122,6 +123,55 @@ func (m *InstancesManager) Resync(ctx context.Context) time.Time {
 	return resyncStart
 }
 
+func (m *InstancesManager) SupportInstanceSync() bool {
+	return true
+}
+
+func (m *InstancesManager) InstanceSync(ctx context.Context, instanceID string) time.Time {
+	resyncStart := time.Now()
+
+	vpcs, err := m.api.GetVPCs(ctx)
+	if err != nil {
+		log.WithError(err).Warning("Unable to synchronize VPC list")
+		return time.Time{}
+	}
+
+	vSwitches, err := m.api.GetVSwitches(ctx)
+	if err != nil {
+		log.WithError(err).Warning("Unable to retrieve VPC vSwitches list")
+		return time.Time{}
+	}
+
+	securityGroups, err := m.api.GetSecurityGroups(ctx)
+	if err != nil {
+		log.WithError(err).Warning("Unable to retrieve ECS security group list")
+		return time.Time{}
+	}
+
+	oldInstances := m.instances.DeepCopy()
+	instances, err := m.api.UpdateInstance(ctx, vpcs, vSwitches, instanceID, oldInstances)
```

Comment:
Two questions here:

* What is the implication of merging maps, could this result in leaking of old instances that may have been removed via the GetInstances call?

* Similar to my other question - this function looks almost identical to Resync, except for using UpdateInstance instead of GetInstance. I'm wondering if we could reduce the duplication a bit?

Reply:
Thanks for the comments! I'll try to address them and refactor the code.

> What is the implication of merging maps, could this result in leaking of old instances that may have been removed via the GetInstances call?
Thanks for pointing this out. One issue I can imagine is that after this change, if an ENI on the instance is somehow detached while maintaining the IP pool, it will not be deleted from the map until the next background full resync, I think this could be fixed, will do that in the next amending.  
Other than that, I think ENIs on different nodes would not overlap and the introduced `instancesAPIResyncLock` would avoid data races.